TaskHandler returns the task_id if persistent

davewiththenicehat opened the pull request:
#### Brief overview of PR changes/additions
TaskHandler returns the task_id if the persistent kwa is set.

#### Motivation for adding to Evennia
https://github.com/evennia/evennia/issues/2370#issuecomment-814948489

#### Other info (issues closed, discussion etc)
Checked over TaskHandler. Am I understanding correct that the task id does not change over a restart. All that is needed for this issue is returning the task_id if the persistent kwa is set. deferLater is called before returning the task_id.
Current unit tests pass with this change.

davewiththenicehat commented:
@Griatch would this be a working solution for 2370?

I was unable to find a unit test for delay or TaskHandler. Was I looking in the wrong locations?

Griatch commented:
On one hand, this will break existing implementations that depend on the deferred, but on the other hand one can argue that without it, it's not really possible to prematurely stop a task from the handler across reloads. 
Unfortunately there doesn't appear to exist any unit tests for TaskHandler; it's something that should be added when we start making changes to it. :/ 

If you are interested in starting a new unit test suit (at least covering this changed return) for the TaskHandler, let me know if you want to do so as part of this PR. Otherwise I'll need to do it before I merge this.

davewiththenicehat commented:
I will add it in to free you up for more complicated tasks.

davewiththenicehat commented:
After a bit of reading up on twisted's reactor and task process I was able to find I needed to create local reference to the clock attribute on TaskHandler.
This allows for the proper creation of a unit test.

I am going to move forward with properly canceling a twisted task within TaskHandler. Allowing users to call TaskHandler methods directly. As TaskHandler currently stands (and previously) if do_task was called directly it creates an AlreadyCalled error when the delay time has completed.
If for any reason this is not wanted let me know. The docstring for TaskHandler suggested this should be possible.

Griatch commented:
Ok, will be interesting to see the change. :+1:

davewiththenicehat commented:
@Griatch @kovitikus 
I completed the changes I was interested in.
TaskHandler.add now returns a Task object. A Task object has the most common methods of a twisted deferred object. Enough to gain full compatibility with existing evennia code, including contribs. In addition it contains forwarding methods to newly created methods. Removing the need to import TASK_HANDLER. It is still possible, just no longer needed.
I've suppressed the cancel error that deferred raises if the deferred is canceled before the callback or errback was called.
If task's are added than canceled but never called task handler will remove these stale tasks after TaskHandler.stale_timeout (default 60) seconds. If TaskHandler.stale_timeout is 0 or less stale tasks are never automatically removed. Because this was not important enough for a timer the check is done after new tasks are added and when TaskHandler.load is called.
Method remove is now functional.
Added the ability to call the callback of a task at any time without affecting the task or it's deferred.
Added methods exists, active to check task states.
TaskHandler is more state aware. It now expects it's tasks to not exist. For example in the add method it is possible for a callback to complete before the completion of twisted's deferLater method. Removing the task from task handler's dictionaries. I added these checks through the module.

Most importantly, I added unit tests. Including mimicking a server restart.
I made certain to use a test function that mocks a real example of a persistent task.
I've run all evennia unit tests. All tested good.

While I have this fresh in memory let me know if there are any other feature recommendations.
Please check my changes.

Had a good bit of fun updating this.

Griatch commented on evennia/utils/tests/test_utils.py:
```diff
@@ -292,3 +295,191 @@ def test_byte_string(self):
+    Test utils.delay.
+    """
+
+    def test_delay(self):
```
Wow ... this is ... very hard to read :sweat_smile: We have some existing tests in Evennia that is structured like a wall of text like this and we are trying to get away from that :wink:. 

I would suggest splitting this into several methods instead, using a `setUp` (and optional `tearDown` method) to set up the taskhandler when testing one specific thing. That way you can run individual tests and it becomes much easier to see which is failing (and you can also have proper doc strings) with less risk of 'contamination' between the tests (each unit test is in principle meant to be stand-alone from others). 
It's still okay to have multiple asserts per test etc, but try to separate the tests into what is being tested.

davewiththenicehat replied:
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

Griatch commented on evennia/scripts/taskhandler.py:
```diff
@@ -101,23 +147,45 @@ def add(self, timedelay, callback, *args, **kwargs):
             any (any): additional keyword arguments to send to the callback
 
+        Returns:
+            task_id (int), the task's id intended for use with this class.
+            False, if the task has completed before addition finishes.
```
This should be written as 

```
Returns:
    int or Task: For non-persistent tasks, the `int` will be the task ID [...]. For persistent tasks, the 
    special `Task` object is returned. This wraps a Twisted deferred and has a similar API. 
```

davewiththenicehat replied:
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

Griatch commented on evennia/scripts/taskhandler.py:
```diff
@@ -101,23 +147,45 @@ def add(self, timedelay, callback, *args, **kwargs):
+
+        Notes:
+            This method has two return types.
+            An instance of twisted's Deferred class is standard.
+            If the persistent kwarg is truthy instead a task ID will be returned.
+            This task id can be used with task handler's do_task and remove methods.
+
```
This is not a `Deferred` anymore, but a `Task`, no?

Griatch commented on evennia/scripts/taskhandler.py:
```diff
@@ -194,10 +446,198 @@ def create_delays(self):
 TASK_HANDLER = TaskHandler()
+
+
+class Task:
```
I worry that the name `Task` is too generic a name for this class. It's a special helper class for use in the `TaskHandler`, but with such a universal name one may think it's more universal. It would make more sense to rename this `TaskHandlerTask` I think.

davewiththenicehat replied:
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

Griatch reviewed (changes requested):
Thanks for the PR! Overall this looks good, and will be a good additional functionality/fix. Kudos for expanding with new unit tests too :+1: 
With a long PR like this, I need to review it pretty carefully, so don't be discouraged over the amount of comments. This just needs a little more 'massaging' before it's ready to go into core. :)

davewiththenicehat commented:
@Griatch addressed issues found.
Be certain the double check each. There was a number of them and I went through them rather fast. It is very possible I missed something.

I did find a minor bug after updating the unit tests and resolved it. The persistent kwarg was not getting removed from the kwarg dict if persistent was pass with a false value. Causing it to remain in the dict and was passed to the callback when it should not have been.

If mergeable, before merging.
Should remove delete the instance of the canceled deferred? Currently it does not.

kovitikus reviewed (changes requested):
Just adding some minor fixes here.

There's a few lines that wouldn't let me comment on them.

On lines `439`, `465`, `531`, and `534` the words `deferral` should be changed to `deferred`.

![line 439](https://i.imgur.com/KSL0w21.png)
![line 465](https://i.imgur.com/AbzSWPq.png)
![lines 531 and 534](https://i.imgur.com/1Cih52x.png)

On line `189`, `...canceled from...` should include `tasks` as `...canceled tasks from...`

![line 189](https://i.imgur.com/1Vj6LP5.png)

On line `194`, `It` is misspelled as `I`, missing its `t`.

![line 194](https://i.imgur.com/eq3d1jP.png)

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -15,20 +15,173 @@
+        deferred (deferred): a reference to this task's deferred
+    Propert Attributes:
+        paused (bool): check if the deferral of a task has been paused.
+        called(self): A task attribute to check if the deferral of a task has been called.
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -15,20 +15,173 @@
+        cancel(): Stop a task from automatically executing.
+        active(): Check if a task is active (has not been called yet).
+        exists(): Check if a task exists.
+        get_id(): Returns the global id for this task. For use with
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -139,35 +290,30 @@ def save(self):
+        longer accurate.
 
         Args:
             timedelay (int or float): time in sedconds before calling the callback.
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -139,35 +290,30 @@ def save(self):
-            As those memory references will no longer acurately point to
-            the variable desired.
+            TaskHandlerTask: An object to represent a task.
+                Reference evennia.scripts.taskhandler.Task for complete details.
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -235,73 +381,56 @@ def add(self, timedelay, callback, *args, **kwargs):
         Returns:
-            True (bool): if the task exists.
-            False (bool): if the task does not exist.
+            bool: Tru the task exists False if it does not.
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -144,47 +212,231 @@ def add(self, timedelay, callback, *args, **kwargs):
+
+        Returns:
+            True (bool): if the task exists.
+            False (bool): if the task does not exist.
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -144,47 +212,231 @@ def add(self, timedelay, callback, *args, **kwargs):
+            del d
+        return True
+
+    def remove_all(self, save=True, cancel=True):
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -58,16 +76,44 @@ def load(self):
 
                 callback = getattr(obj, method)
-            self.tasks[task_id] = (date, callback, args, kwargs)
+            self.tasks[task_id] = date, callback, args, kwargs, True, None
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

davewiththenicehat commented on evennia/scripts/taskhandler.py:
```diff
@@ -6,31 +6,49 @@
         self.to_save = {}
+        self.clock = reactor
+        # number of seconds before an uncalled canceled task is removed from TaskHandler
+        self.stale_timeout = 60
```
Done with https://github.com/evennia/evennia/pull/2374/commits/cf8052b816c0011330ead39c85e871c85918a0e4

kovitikus commented:
Looking AWESOME!!! 😍

Griatch reviewed (approved):
This looks good now, thanks for persevering and getting it all fixed! During testing I found a bug where adding a non-pickleable callable causes all subsequent adds to fail. This is because the faulty callable gets added to internal storage before the callable is actually tested. I will fix this before merging.

Griatch commented:
@davewiththenicehat In the future, consider not doing a change like this in your `master` branch. It's better to do it in a separate work branch that you can then delete once upstream master merges it. That way your local `master` branch in always sync with upstream.
